docs(api): new_pipette.rst revisions for Flex (#13218)
OpenTrons/opentrons-api: api/docs/v2/new_pipette.rst

Loading a Flex 96-Channel Pipette
---------------------------------

This code sample loads the Flex 96-Channel Pipette. Because of its size, the Flex 96-Channel Pipette requires the left *and* right pipette mounts. You cannot use this pipette with 1- or 8-Channel Pipette in the same protocol or when these instruments are attached to the robot. To load the 96-Channel Pipette, specify its position as ``mount='left'`` as shown here:

.. code-block:: python

    def run(protocol: protocol_api.ProtocolContext):
        left = protocol.load_instrument(
            instrument_name='flex_96channel_1000', mount='left')

.. versionadded:: 2.15

Loading OT-2 Pipettes
---------------------

This code sample loads a P1000 Single-Channel GEN2 pipette in the left mount and a P300 Single-Channel GEN2 pipette in the right mount. Each pipette uses its own 1000 µL tip rack. 

.. code-block:: python
    :substitutions:

    from opentrons import protocol_api

>     metadata = {'apiLevel': |apiLevel|}

Reviewer: need the quotes here. the value has to be a string, or it raises an error.

Multi-Channel Pipettes
======================

All building block and advanced commands work with single- and multi-channel pipettes.

To keep the interface to the Opentrons API consistent between single- and multi-channel pipettes, commands treat the *backmost channel* (furthest from the door) of a multi-channel pipette as the location of the pipette. Location arguments to building block and advanced commands are specified for the backmost channel.

Also, this means that offset changes (such as :py:meth:`.Well.top` or :py:meth:`.Well.bottom`) can be applied to the single specified well, and each pipette channel will be at the same position relative to the well that it is over.

Finally, because there is only one motor in a multi-channel pipette, these pipettes always aspirate and dispense on all channels simultaneously.

8-Channel, 96-Well Plate Example
--------------------------------

To demonstrate these concepts, let's write a protocol that uses a Flex 8-Channel Pipette and a 96-well plate. We'll then aspirate and dispense a liquid to different locations on the same well plate. To start, let's load a pipette in the right mount and add our labware.

> .. code-block:: python

Reviewer: maybe include `requirements` here. the previous example is OT-2.

After picking up a tip, let's tell the robot to aspirate 100 µL from the well plate at location ``A1``::

    right.aspirate(volume=100, location=plate['A1'])

The eight pipette channels will only aspirate from every other well in the column: A1, C1, E1, G1, I1, K1, M1, and O1.

Finally, let's tell the robot to dispense 100 µL into the well plate at location ``B1``::

    right.dispense(volume=100, location=plate['B1'])

The eight pipette channels will only dispense into every other well in the column: B1, D1, F1, H1, J1, L1, N1, and P1.

Adding Tip Racks
================

The ``load_instrument()`` method includes the optional argument ``tip_racks``. This parameter accepts a list of tip rack labware objects, which lets you to specify as many tip racks as you want. The advantage of using ``tip_racks`` is twofold. First, associating tip racks with your pipette allows for automatic tip tracking throughout your protocol. Second, it removes the need to specify tip locations in the :py:meth:`.InstrumentContext.pick_up_tip` method. For example, let's start by loading loading some labware and instruments like this::
        
    def run(protocol: protocol_api.ProtocolContext):

>     tiprack_left = protocol.load_labware(

Reviewer: \+ indent level for everything under `def run`

            load_name='corning_96_wellplate_360ul_flat', location='D3')
        pipette.pick_up_tip()

Let's tell the robot to aspirate, dispense, and blow out the liquid using default flow rates. Notice how you don't need to specify a ``flow_rate`` attribute to use the defaults::

        pipette.aspirate(200, plate['A1'])  # 160 µL/s
        pipette.dispense(200, plate['A2'])  # 160 µL/s
        pipette.blow_out()                  #  80 µL/s

Now let's change the flow rates for each action::

        pipette.flow_rate.aspirate = 50
        pipette.flow_rate.dispense = 100
        pipette.flow_rate.blow_out = 75
        pipette.aspirate(200, plate['A1'])  #  50 µL/s
        pipette.dispense(200, plate['A2'])  # 100 µL/s
        pipette.blow_out()                  #  75 µL/s

.. note::
    In API version 2.13 and earlier, :py:obj:`.InstrumentContext.speed` offered similar functionality. It attempted to set the plunger speed in mm/s. Due to technical limitations, that speed could only be approximate. You must use ``.flow_rate`` in version 2.14 and later, and you should consider replacing older code that sets ``.speed``.

.. versionadded:: 2.0

> .. _pipette:

Reviewer: this is a very generic anchor name. it doesn't seem that we're linking here from anywhere, so we can remove it.

Pipettes
########

When writing a protocol, you must inform the Protocol API about the pipettes you will be using on your robot. The :py:meth:`.ProtocolContext.load_instrument` function provides this information and returns an :py:class:`.InstrumentContext` object.

For information about liquid handling, see :ref:`v2-atomic-commands` and :ref:`v2-complex-commands`.

.. _new-create-pipette:

Loading Pipettes
================

You load pipettes in a protocol using the :py:meth:`~.ProtocolContext.load_instrument` method. This method requires the :ref:`pipette's API load name <new-pipette-models>`, its left or right mount position, and (optionally) a list of associated tip racks.

> Similar to working with labware and modules, you must inform the robot about the pipettes you want to use in your protocol. Even if you don't use the pipette anywhere else in your protocol, the Opentrons App and the robot won't let you start the protocol run until all pipettes loaded by ``load_instrument()`` are attached properly.

Reviewer: this language, "you must inform the robot about the pipettes you want to use in your protocol", repeats the opening sentence too closely. the rest of the info in this ¶ is good.

Author: Removing that paragraph and incorporating into 2nd paragraph of 1st section. 
 
> Similar to working with labware and modules, you must inform the robot about the pipettes you want to use in your protocol. As noted above, the:py:meth:`~.ProtocolContext.load_instrument` method provides this capability. It also requires the :ref:`pipette's API load name <new-pipette-models>`, its left or right mount position, and (optionally) a list of associated tip racks. Even if you don't use the pipette anywhere else in your protocol, the Opentrons App and the robot won't let you start the protocol run until all pipettes loaded by ``load_instrument()`` are attached properly.

Loading Flex 1- and 8-Channel Pipettes
--------------------------------------

> This code sample loads a Flex 8-Channel Pipette in the left slot along with a Flex 1-Channel Pipette and two tip racks in the right slot. Both pipettes are 1000 µL. 

Reviewer: makes it sound like the tip racks are in the right pipette mount! either remove that detail, or move it to a second sentence.

also, to avoid confusion, always use _mount_ for pipette locations. _slots_ are on the deck.

Author: Replacing "slot" with "mount."

.. code-block:: Python
    :substitutions:
    
    from opentrons import protocol_api
    
    requirements = {'robotType': 'Flex', 'apiLevel': '|apiLevel|'}

    def run(protocol: protocol_api.ProtocolContext):
        left = protocol.load_instrument(
            instrument_name='flex_8channel_1000', mount='left')   
        tiprack1 = protocol.load_labware(
            load_name='opentrons_flex_96_tiprack_1000ul', location='D1')
        tiprack2 = protocol.load_labware(
            load_name='opentrons_flex_96_tiprack_1000ul', location='C1')
        right = protocol.load_instrument(
            instrument_name='flex_1channel_1000',
            mount='right',

>             tip_racks=[tiprack1, tiprack2]) 

Reviewer: it's a little weird to assign no tip racks to one pipette and two to the other — this isn't best practice, so it's a little weird to neutrally present it as the first example of `load_instrument()`. i think it would be far better to have 1 rack per pipette.

Author: Code samples from original, OT-2 version. Revising to use 1 rack per pipette, changing to load 1 one channel and 8 channel.

Loading a Flex 96-Channel Pipette
---------------------------------

This code sample loads the Flex 96-Channel Pipette. Because of its size, the Flex 96-Channel Pipette requires the left *and* right pipette mounts. You cannot use this pipette with a 1- or 8-Channel Pipette in the same protocol or when these other instruments are attached to the robot. To load the 96-Channel Pipette, specify its position as ``mount='left'`` as shown here:

.. code-block:: python

    def run(protocol: protocol_api.ProtocolContext):
        left = protocol.load_instrument(
            instrument_name='flex_96channel_1000', mount='left')

> .. version added, 2.15??

Reviewer: yes, that seems right. `flex_96channel_1000` is a valid load name in 2.15.

Author: Added version.

Loading OT-2 Pipettes
---------------------

> This code sample loads a 8-channel, P300 GEN2 pipette in the left slot and a single-channel, P1000 GEN2 pipette with two tip racks in the right slot. 

Reviewer: use standard product name order:
- P300 8-Channel GEN2
- P100 Single-Channel GEN2

Author: Replacing with correct product names.

- Flex 1-Channel Pipette
- Flex 8-Channel Pipette

.. code-block:: python

    from opentrons import protocol_api

>     metadata = {'apiLevel': '2.14'}

Reviewer: nothing stopping this from being `|apiLevel|` that i know of

Author: Revised as:

> metadata = {'apiLevel': |apiLevel|}

    def run(protocol: protocol_api.ProtocolContext):
        left = protocol.load_instrument(
            instrument_name='p300_multi_gen2',
            mount='left')
        tiprack1 = protocol.load_labware(
            load_name='opentrons_96_tiprack_1000ul',
            location=1)
        tiprack2 = protocol.load_labware(
            load_name='opentrons_96_tiprack_1000ul',
            location=2)
        right = protocol.load_instrument(
            instrument_name='p1000_single_gen2',
            mount='right',

>             tip_racks=[tiprack1, tiprack2])

Reviewer: mirror any change to tip rack loading in Flex example above.

Author: These have changed and mirror each other.

- loading tip racks 1st
- loading instruments after

.. versionadded:: 2.0

.. _new-multichannel-pipettes:

Multi-Channel Pipettes
======================

All building block and advanced commands work with single- and multi-channel pipettes.

To keep the interface to the Opentrons API consistent between single- and
multi-channel pipettes, commands treat the *backmost channel* (furthest from the
door) of a multi-channel pipette as the location of the pipette. Location arguments to
building block and advanced commands are specified for the backmost channel.

This also means that offset changes (such as :py:meth:`.Well.top` or
:py:meth:`.Well.bottom`) can be applied to the single specified well, and each
pipette channel will be at the same position relative to the well
that it is over.

> Because there is only one motor in a multi-channel pipette, these pipettes always aspirate and dispense on all channels simultaneously. For instance, to aspirate from the first column of a 96-well plate you would write:
> 
> .. code-block:: python

Reviewer: ```suggestion
Because there is only one motor in a multi-channel pipette, these pipettes always aspirate and dispense on all channels simultaneously. For instance, to aspirate from the first column of a 96-well plate::
```
i'm not a big fan of the "you would write" type constructions. i prefer the concision of "To do X: code for X".

Author: The example after this text, I've separated into two H3 sections. Seem to work better as there's a lot of info being conveyed here. New outline is:

- H2 Multi-Channel Pipettes and Well Plates
- H3 8-Channel, 96-Well Plate Example
- H3 8-Channel, 384-Well Plate Example

This was also the big code sample including extensive multi-line comments. Broke it up. Took comments out of code and incorporated into text with short code examples.

    def run(protocol: protocol_api.ProtocolContext):

>         # Load a tiprack for 1000uL tips

Reviewer: ```suggestion
        # Load a tip rack for 1000 µL tips
```
let's clean these up to match house style

        tiprack1 = protocol.load_labware(
        load_name='opentrons_flex_96_tiprack_1000ul',
        location='D1')

>         # Load a wellplate

Reviewer: ```suggestion
        # Load a well plate
```
here too

Author: Fixed throughout.

Loading Pipettes
================

> Similar to working with labware and modules, you must inform the robot about the pipettes you want to use in your protocol. As noted above, the:py:meth:`~.ProtocolContext.load_instrument` method provides this capability. It also requires the :ref:`pipette's API load name <new-pipette-models>`, its left or right mount position, and (optionally) a list of associated tip racks. Even if you don't use the pipette anywhere else in your protocol, the Opentrons App and the robot won't let you start the protocol run until all pipettes loaded by ``load_instrument()`` are attached properly.

Reviewer: ```suggestion
Similar to working with labware and modules, you must inform the robot about the pipettes you want to use in your protocol. As noted above, the :py:meth:`~.ProtocolContext.load_instrument` method provides this capability. It also requires the :ref:`pipette's API load name <new-pipette-models>`, its left or right mount position, and (optionally) a list of associated tip racks. Even if you don't use the pipette anywhere else in your protocol, the Opentrons App and the robot won't let you start the protocol run until all pipettes loaded by ``load_instrument()`` are attached properly.
```
syntax fix

.. code-block:: python

    def run(protocol: protocol_api.ProtocolContext):
        left = protocol.load_instrument(
            instrument_name='flex_96channel_1000', mount='left')

.. versionadded:: 2.15

Loading OT-2 Pipettes
---------------------

This code sample loads a P1000 Single-Channel GEN2 pipette in the left mount and a P300 Single-Channel GEN2 pipette in the right mount. Each pipette uses its own 1000 µL tip rack. 

.. code-block:: python
    :substitutions:

    from opentrons import protocol_api

    metadata = {'apiLevel': |apiLevel|}

    def run(protocol: protocol_api.ProtocolContext):
        tiprack1 = protocol.load_labware(
            load_name='opentrons_96_tiprack_1000ul',
            location=1)
        tiprack2 = protocol.load_labware(
            load_name='opentrons_96_tiprack_1000ul',
            location=2)
        left = protocol.load_instrument(
            instrument_name='p1000_single_gen2',
            mount='left',
            tip_racks=[tiprack1])
        right = protocol.load_instrument(
            instrument_name='p300_multi_gen2',
            mount='right',
            tip_racks=[tiprack1])

.. versionadded:: 2.0

> When you load a pipette in this way, you are declaring that you want the specified pipette to be attached to the robot. Even if you don't use the pipette anywhere else in your protocol, the Opentrons App or the touchscreen on Flex will not let your protocol proceed until all pipettes loaded with ``load_instrument`` are attached.

Reviewer: can delete, this is covered in the intro.

Author: oops. fat finger. Fixing this locally. Will change and push.

> If you're writing a protocol that uses the Flex Gripper, you might think that this would be the place in your protocol to declare that. However, the gripper doesn't require ``load_instrument``! Whether your gripper requires a protocol is determined by the presence of :py:meth:`.ProtocolContext.move_labware` commands. See :ref:`moving-labware` for more details.

Reviewer: should go up in the Flex section

Author: Yikes. Yes it should. Fixing locally and will push.

.. _new-multichannel-pipettes:

> Multi-Channel Pipettes and Well Plates

Reviewer: check the cross-references to make sure this won't add the text "and Well Plates" to links elsewhere.

Author: Let's go back to the original, "Multi-Channel Pipettes."  I didn't really like the extra I added on -- too long.

Also, this means that offset changes (such as :py:meth:`.Well.top` or :py:meth:`.Well.bottom`) can be applied to the single specified well, and each pipette channel will be at the same position relative to the well that it is over.

Finally, because there is only one motor in a multi-channel pipette, these pipettes always aspirate and dispense on all channels simultaneously.

8-Channel, 96-Well Plate Example
--------------------------------

To demonstrate these concepts, let's write a protocol that uses a Flex 8-Channel Pipette and a 96-well plate. We'll then aspirate and dispense a liquid to different locations on the same well plate. To start, let's load a pipette in the right mount and add our labware.

.. code-block:: python

    def run(protocol: protocol_api.ProtocolContext):
        # Load a tiprack for 1000uL tips
        tiprack1 = protocol.load_labware(
        load_name='opentrons_flex_96_tiprack_1000ul',

>         location='D1')       

Reviewer: fix indents here, Python will consider this a syntax error

Author: 2x yikes. Fixing locally. Will push revisions.

        # Load a 96-well plate
        plate = protocol.load_labware(
            load_name='corning_96_wellplate_360ul_flat',
            location='C1')       
        # Load an 8-channel pipette on the right mount
        right = protocol.load_instrument(
            instrument_name='flex_8channel_1000',
            mount='right',
            tip_racks=[tiprack1])

After loading our instruments and labware, let's tell the robot to pick up a pipette tip from location ``A1`` in ``tiprack1``::

>     right.pick_up_tip(tiprack1['A1'])

Reviewer: the code sets up tip tracking for this pipette. so i'd expect just `right.pick_up_tip()`. 

if you want to be explicit about the location on the tip rack, leave this line as is and remove the `tip_racks` parameter from `right = protocol.load_instrument(...`

Author: Hey, thanks for pointing that out. I had a TIL moment with this further down in the flow rate section. Let's go with `right.pick_up_tip()` because we've got that specified already.

With the backmost pipette channel above location A1 on the tip rack, all eight channels are above the eight tip rack wells in column 1.   

> After picking up a tip, let's tell the robot to aspirate 300 ul from the well plate at location ``A2``::

Reviewer: ```suggestion
After picking up a tip, let's tell the robot to aspirate 300 µL from the well plate at location ``A2``::
```
unit style. there's a few of these to find and replace.

Author: Doing a search and replace now.

        
    right.aspirate(volume=300, location=plate['A2'])

> With the backmost pipette tip above location A2 on the well plate, all eight channels are above the eight wells in column 1.

Reviewer: ```suggestion
With the backmost pipette tip above location A2 on the well plate, all eight channels are above the eight wells in column 2.
```

Author: Fixing locally. Will push.

Finally, let's tell the robot to dispense 300 ul into the well plate at location ``A3``::

    right.dispense(volume=300, location=plate['A3'].top())

With the backmost pipette tip above location A3, all 8 channels are above the eight wells in column 3. The pipette will dispense liquid into all the wells simultaneously.

8-Channel, 384-Well Plate Example
---------------------------------

> In general, you should specify wells in the first row of a well plate when using multi-channel pipettes. An exception to this rule is when using 384-well plates. The greater well density means the nozzles of a multi-channel pipette can only accesses every other well in a column. Specifying well A1 accesses every other well starting with the first (rows A, C, E, G, I, K, M, and O). Similarly, specifying well B1 also accesses every other well, but starts with the second (rows B, D, F, H, J, L, N, and P).

Reviewer: ```suggestion
In general, you should specify wells in the first row of a well plate when using multi-channel pipettes. An exception to this rule is when using 384-well plates. The greater well density means the nozzles of a multi-channel pipette can only access every other well in a column. Specifying well A1 accesses every other well starting with the first (rows A, C, E, G, I, K, M, and O). Similarly, specifying well B1 also accesses every other well, but starts with the second (rows B, D, F, H, J, L, N, and P).
```
`can only access` or `only accesses`

Author: My god, I am illiterate. Fixing locally. Will commit.

To demonstrate these concepts, let's write a protocol that uses a Flex 8-Channel Pipette and a 384-well plate. We'll then aspirate and dispense a liquid to different locations on the same well plate. To start, let's load a pipette in the right mount and add our labware.

.. code-block:: python

    def run(protocol: protocol_api.ProtocolContext):
        # Load a tiprack for 200uL tips
        tiprack1 = protocol.load_labware(
            load_name='opentrons_flex_96_tiprack_200ul', location=1)
        # Load a well plate
        plate = protocol.load_labware(
            load_name='corning_384_wellplate_112ul_flat', location=4)
        # Load an 8-channel pipette on the right mount
        right = protocol.load_instrument(
            instrument_name='flex_8channel_1000',
            location='right',
            tip_racks=[tiprack1])

After loading our instruments and labware, let's tell the robot to pick up a pipette tip from location ``A1`` in ``tiprack1``::

>     right.pick_up_tip(tiprack1['A1'])

Reviewer: make same changes to tip tracking as in 96-channel example

With the backmost pipette channel above location A1 on the tip rack, all eight channels are above the eight tip rack wells in column 1.

After picking up a tip, let's tell the robot to aspirate 100 ul from the well plate at location ``A1``::

    right.aspirate(volume=100, location=plate['A1'])

> Because of the limited clearance between wells, the eight pipette channels will only aspirate from every other well in the column (e.g. A1, C1, E1, G1, I1, K1, M1, O1).

Reviewer: ```suggestion
The eight pipette channels will only aspirate from every other well in the column: A1, C1, E1, G1, I1, K1, M1, O1.
```
text above explains this well enough, i think

Author: Fixing locally. Will commit.

Finally, let's tell the robot to dispense 100 ul into the well plate at location ``B1``::

    right.dispense(volume=100, location=plate['B1'])

> Because of the limited clearance between wells, the eight pipette channels will only dispense into every other well in the column (e.g. B1, D1, F1, H1, J1, L1, N1, P1).

Reviewer: simplify in same way as line 179

Author: Fixing locally.

Adding Tip Racks
================

The ``load_instrument()`` method includes the optional argument ``tip_racks``. This parameter accepts a list of tip rack labware objects, which lets you to specify as many tip racks as you want. The advantage of using ``tip_racks`` is twofold. First, associating tip racks with your pipette allows for automatic tip tracking throughout your protocol. Second, it removes the need to specify tip locations in the :py:meth:`.InstrumentContext.pick_up_tip` method. For example, let's start by loading loading some labware and instruments like this::
        
    def run(protocol: protocol_api.ProtocolContext):
    tiprack_left = protocol.load_labware(

>         load_name='opentrons_flex_96_tiprack_200ul', location='1')

Reviewer: Flex labware, OT-2 deck slot. (nice that it doesn't matter!)

Author: Found 2 locations. Fixing slots 1 & 2 to D1 & D2.

    tiprack_right = protocol.load_labware(
        load_name='opentrons_flex_96_tiprack_200ul', location='2')
    left_pipette = protocol.load_instrument(
        instrument_name='flex_8channel_1000', mount='left')
    right_pipette = protocol.load_instrument(
            instrument_name='flex_8channel_1000',
            mount='right',
            tip_racks=[tiprack_right])

> Next, let's specify the tip rack location for the left pipette, which was loaded without a ``tip_racks`` argument::

Reviewer: This explanation makes it sound like the `pick_up_tip` command will associate the tip rack with the pipette for the future, not just for that one action. Maybe something like `Let's pick up a tip with the left pipette. We need to specify the location as an argument of ``pick_up_tip()``, since we loaded the left pipette without a ``tip_racks`` argument.`

Author: Great recommendation. Fixing.

    
    left_pipette.pick_up_tip(tiprack_left['A1'])
    left_pipette.drop_tip()

But now you have to specify ``tiprack_left`` every time you call ``pick_up_tip``, which means you're doing all your own tip tracking::

    left_pipette.pick_up_tip(tiprack_left['A2'])
    left_pipette.drop_tip()
    left_pipette.pick_up_tip(tiprack_left['A3'])
    left_pipette.drop_tip()

However, because you specified a tip rack location for the right pipette, the robot will automatically pick up from location ``A1`` of its associated tiprack::
    
    right_pipette.pick_up_tip()
    right_pipette.drop_tip()

Additional calls to ``pick_up_tip`` will automatically progress through the tips in the right rack::

> 

Reviewer: the below code block is a great place to use inline comments to show the effect of a command. always put two spaces before the comment.

```
right_pipette.pick_up_tip()  # picks up from A2
right_pipette.drop_tip()
right_pipette.pick_up_tip()  # picks up from A3
right_pipette.drop_tip()
```

Author: Fixing locally. Taking note on comment spacing too.

        # Similarly, specifying well A2 for aspirate means the pipette will
        # position its backmost channel over well A2, and the rest of the
        # pipette channels are over the rest of the wells of column 1
        right.aspirate(volume=300, location=plate['A2'])

        # Dispense into column 3 of the plate with all 8 channels of the
        # pipette at the top of their respective wells

>         right.dispense(volume=300, location=plate['A3'].top())

Reviewer: These long code comments are hard to read. I'd rather break up the snippet than make people read dark grey monospace on light grey.

> In general, you should specify wells in the first row of a well plate when using multi-channel pipettes. One exception to this rule is when using 384-well plates. The limited space between the wells in a 384-well plate and between the nozzles of a multi-channel pipette means the pipette accesses every other well in a column. Specifying well A1 accesses every other well starting with the first (rows A, C, E, G, I, K, M, and O). Similarly, specifying well B1 also accesses every other well, but starts with the second (rows B, D, F, H, J, L, N, and P).

Reviewer: it's not that the space is "limited" — kind of the opposite. 384 plates have double the density (horizontal and vertical) of 96 plates.

Author: Revised to remove "limited."

> The greater well density means the nozzles of a multi-channel pipette can only accesses every other well in a column.

.. code-block:: python

    def run(protocol: protocol_api.ProtocolContext):
        # Load a tiprack for 200uL tips
        tiprack1 = protocol.load_labware(
            load_name='opentrons_flex_96_tiprack_200ul', location=1)
        # Load a wellplate
        plate = protocol.load_labware(
            load_name='corning_384_wellplate_112ul_flat', location=4)

>         # Load an 8-channel Multi GEN2 on the right mount

Reviewer: ```suggestion
        # Load a P300 8-Channel GEN2 on the right mount
```

Author: This is fixed in a recent draft push.

        right = protocol.load_instrument(
            instrument_name='flex_8channel_1000',
            location='right',
            tip_racks=[tiprack1])

        # pick up a tip in preparation for aspiration
        right.pick_up_tip()

        # Aspirate from wells A1, C1, E1, G1, I1, K1, M1, and O1

>         right.aspirate(volume=100, location=plate['A1'])

Reviewer: good catch! 300 µL doesn't fit in a 384-well plate.

        # Dispense in wells B1, D1, F1, H1, J1, L1, N1, and P1
        right.dispense(volume=100, location=plate['B1'])

This pattern of access applies to both building block commands and advanced
commands.

.. _new-pipette-models:

API Load Names
==============

The pipette's API load name (``instrument_name``) is the first parameter of the ``load_instrument()`` method. It tells your robot which attached pipette you're going to use in a protocol. The tables below list the API load names for the currently available Flex and OT-2 pipettes.

.. tabs::

    .. tab:: Flex Pipettes
        
        +-------------------------+-----------+-------------------------+
        | Pipette Name            | Capacity  | API Load Name           |
        +=========================+===========+=========================+
        | Flex 1-Channel Pipette  | 0.5–50 µL | ``flex_1channel_50``    |
        +                         +-----------+-------------------------+
        |                         | 5–1000 µL | ``flex_1channel_1000``  |
        +-------------------------+-----------+-------------------------+

>         | Flex 8-Channel Pipette  + 0.5–50 µL + ``flex_8channel_50``    |

Reviewer: min volume of smaller Flex pipettes is 1 µL, not 0.5

Author: Fixing minimum volume in next push.  Change to 1-50 uL

        +-------------------------+-----------+-------------------------+
        | Flex 96-Channel Pipette | 5–1000 µL | ``flex_96channel_1000`` |
        +-------------------------+-----------+-------------------------+

    .. tab:: OT-2 Pipettes

        +-----------------------------+--------------------+-----------------------+
        | Pipette Name                | Capacity           | API Load Name         |
        +=============================+====================+=======================+
        | P20 GEN2 (single channel)   | 1-20 µL            | ``p20_single_gen2``   |
        +-----------------------------+                    +-----------------------+
        | P20 GEN2 (8-channel)        |                    | ``p20_multi_gen2``    |
        +-----------------------------+--------------------+-----------------------+
        | P300 GEN2 (single chanel)   | 20-300 µL          | ``p300_single_gen2``  |
        +-----------------------------+                    +-----------------------+
        | P300 GEN2 (8-channel)       |                    | ``p300_multi_gen2``   |
        +-----------------------------+--------------------+-----------------------+

>         | P1000 GEN2 (single channel) | 100-1000 µL        | ``p1000_single_gen2`` |

Reviewer: reorder, capitalize, and hyphenate names: e.g. P20 Single-Channel GEN2

Author: Fixed to use proper pipette names e.g.

- P20 Single-Channel GEN2
- Flex 1-Channel Pipette

        +-----------------------------+--------------------+-----------------------+

        See the OT-2 Pipette Generations section below if you're using GEN1 pipettes on an OT-2. The GEN1 family includes the P10, P50, and P300 single- and multi-channel pipettes, along with the P1000 single-chanel model.

OT-2 Pipette Generations
========================

The OT-2 works with the GEN1 and GEN2 pipette models. The newer GEN2 pipettes have different volume ranges than the older GEN1 pipettes. With some exceptions, the volume ranges for GEN2 pipettes overlap those used by the GEN1 models. If your protocol specifies a GEN1 pipette, but you have a GEN2 pipette with a compatible volume range, you can still run your protocol. The OT-2 will consider the GEN2 pipette to have the same minimum volume as the GEN1 pipette. The following table lists the volume compatibility between the GEN2 and GEN1 pipettes.

> .. list-table::

Reviewer: as i read the formatted table, i find that my eye glances right past the volume figures. what if we had three columns: GEN2 Pipette, GEN1 Pipette, GEN1 min. vol.?

the first row then would say P20 Single GEN2, P10 Single GEN1, 1 µL

Author: Fixed to show GEN2 name, GEN1 name, GEN1 volume.

    :header-rows: 1
    
    * - GEN2 Pipette
      - GEN1 Pipette
    * - P20 Single GEN2 (1-20 µL)
      - P10 Single GEN1 (1-10 µL)
    * - P20 Multi-Channel GEN2 (1-20 µL)
      - P10 Multi GEN1 (1-10 µL)
    * - P300 Single GEN2 (20-300 µL)
      - P300 Single GEN1 (30-300 µL)
    * - P300 Multi-Channel GEN2 (20-300 µL)
      - P300 Multi GEN1 (20-200 µL)
    * - P1000 Single GEN2 (100-1000 µL)
      - P1000 Single GEN1 (100-1000 µL)

The single- and multi-channel P50 GEN1 pipettes are the exceptions here. If your protocol uses a P50 GEN1 pipette, there is no backward compatibility with a related GEN2 pipette. To replace a P50 GEN1 with a corresponding GEN2 pipette, edit your protocol to load a P20 Single GEN2 (for volumes below 20 µL) or a P300 Single GEN2 (for volumes between 20 and 50 µL).

Adding Tip Racks
================

> The ``load_instrument()`` method includes the optional argument, ``tip_racks``. This parameter accepts a list of tip rack labware objects, which lets you to specify as many tip racks as you want. The advantage of using ``tip_racks`` is twofold. First, associating tip racks with your pipette allows for automatic tip tracking throughout your protocol. Second, it removes the need to specify tip locations in the :py:meth:`.InstrumentContext.pick_up_tip` method. For example::

Reviewer: ```suggestion
The ``load_instrument()`` method includes the optional argument ``tip_racks``. This parameter accepts a list of tip rack labware objects, which lets you to specify as many tip racks as you want. The advantage of using ``tip_racks`` is twofold. First, associating tip racks with your pipette allows for automatic tip tracking throughout your protocol. Second, it removes the need to specify tip locations in the :py:meth:`.InstrumentContext.pick_up_tip` method. For example::
```
not an appositive. i will die on this hill 😆

Author: No need to die. Removing comma.

        
    def run(protocol: protocol_api.ProtocolContext):
    tiprack_left = protocol.load_labware(
        load_name='opentrons_96_tiprack_300ul', location='1')
    tiprack_right = protocol.load_labware(
        load_name='opentrons_96_tiprack_300ul', location='2')
    left_pipette = protocol.load_instrument(
        instrument_name='p300_single', mount='left')
    right_pipette = protocol.load_instrument(
            instrument_name='p300_multi',
            mount='right',
            tip_racks=[tiprack_right])

>     # You must specify the tip location for the left pipette, which was

Reviewer: i'd also like to split these comments out into body paragraphs.

Author: Yes, new changes break up the code and move revised comments to regular text.

    # loaded without specifying tip_racks
    left_pipette.pick_up_tip(tiprack_left['A1'])
    left_pipette.drop_tip()

    # And you have to do it every time you call pick_up_tip, doing all
    # your own tip tracking
    left_pipette.pick_up_tip(tiprack_left['A2'])
    left_pipette.drop_tip()
    left_pipette.pick_up_tip(tiprack_left['A3'])
    left_pipette.drop_tip()

    # Since you specified tip_racks when loading the right pipette, it will
    # automatically pick up from A1 of its associated tiprack
    right_pipette.pick_up_tip()
    right_pipette.drop_tip()

    # And further calls to pick_up_tip will automatically progress through
    # the tips in the rack
    right_pipette.pick_up_tip()
    right_pipette.drop_tip()
    right_pipette.pick_up_tip()
    right_pipette.drop_tip()
       
See also, :ref:`v2-atomic-commands` and :ref:`v2-complex-commands`.

.. versionadded:: 2.0

.. _new-plunger-flow-rates:

Pipette Flow Rates
==================

> Pipettes aspirate or dispense at different rates measured in units of µL/second. You can change the flow rate on a loaded :py:class:`.InstrumentContext` by altering the :py:obj:`~.InstrumentContext.flow_rate` properties listed below:

Reviewer: probably `µL/s` or `microliters per second`. this feels like a hybrid.

Author: Committing in next push.

* Aspirate: ``InstrumentContext.flow_rate.aspirate``
* Dispense: ``InstrumentContext.flow_rate.dispense``
* Blow-out: ``InstrumentContext.flow_rate.blow_out``

These flow rate properties operate independently. This means you can specify different flow rates for each property within the same protocol. For example, let's load a simple protocol and set different flow rates for the attached pipette.

.. code-block:: python

    def run(protocol: protocol_api.ProtocolContext):
        tiprack1 = protocol.load_labware(
            load_name='opentrons_flex_96_tiprack_1000ul',
            location='D1')       
    pipette = protocol.load_instrument(
            instrument_name='flex_1channel_1000',
            mount='left',
            tip_racks=[tiprack1])                
    plate = protocol.load_labware(
            load_name='corning_96_wellplate_360ul_flat',
            location='D3')
    pipette.pick_up_tip()

Let's tell the robot to aspirate, dispense, and blow out the liquid using default flow rates. Notice how you don't need to specify a ``flow_rate`` attribute to use the defaults::

>         pipette.aspirate(200, plate['A1'])

Reviewer: another good spot for comments
```
pipette.aspirate(200, plate['A1'])  # 160 µL/s
pipette.dispense(200, plate['A2'])  # 160 µL/s
pipette.blow_out()                  #  80 µL/s
```
here it makes sense to do some creative spacing with the last one for readability

Author: Fixing locally.

        pipette.dispense(200, plate['A2'])
        pipette.blow_out()

Now let's change the flow rates for each action::

        pipette.flow_rate.aspirate = 50
        pipette.flow_rate.dispense = 100
        pipette.flow_rate.blow_out = 75

>         pipette.aspirate(200, plate['A1'])

Reviewer: like above
```
pipette.aspirate(200, plate['A1'])  #  50 µL/s
pipette.dispense(200, plate['A2'])  # 100 µL/s
pipette.blow_out()                  #  75 µL/s
```

Author: Fixing locally.

        pipette.drop_tip()

.. Docs notes this property was removed in 2.14. If removed, why keep?
:py:obj:`.InstrumentContext.speed` offers the same functionality, but controlled in
units of mm/s of plunger speed. This does not have a linear transfer to flow rate and

> should only be used if you have a specific need.

Reviewer: Docs cover all versions, up through the latest and greatest (2.15). It's totally valid to use older versions on OT-2, even into the future. But this was _removed for a good reason_ so we should probably caution users against it fairly strongly. Something like:

```
.. note::
    In API version 2.13 and earlier, :py:obj:`.InstrumentContext.speed` offered similar functionality. It attempted to set the plunger speed in mm/s. Due to technical limitations, that speed could only be approximate. You must use ``.flow_rate`` in version 2.14 and later, and you should consider replacing older code that sets ``.speed``.
```

Author: Committing in next push.

See :ref:`gantry_speed` and :ref:`axis_speed_limits` for information about other ways of controlling pipette movement.

.. versionadded:: 2.0

.. _defaults:

Flex Pipette Flow Rates
-----------------------

The following table provides data on the default aspirate, dispense, and blow-out flow rates (in µL/s) for Flex pipettes.

INFORMATION TBD 

OT-2 Pipette Flow Rates
-----------------------

The following table provides data on the default aspirate, dispense, and blow-out flow rates (in µL/s) for OT-2 GEN2 pipettes.

+---------------------------------+---------------+---------------+---------------+

> | Pipette                         | Aspirate      | Dispense      | Blow-out      |

Reviewer: I don't think we need the split header cell here. But we may want to split volume into its own column, since it comes at the end of every entry under Pipette. So:
Pipette | Volume range | Aspirate, dispense, blow-out rate

Reviewer: or just "Flow rate" for the third column

Author: Went with "Flow Rates (uL/s)". Including the units in header for extra reinforcement/reminder.

+=================================+===============+===============+===============+ 
| P20 Single GEN2 (1-20 µL)       | * API v2.6 or higher: 7.56 µL/s               |
|                                 | * API v2.5 or lower: 3.78 µL/s                |
+---------------------------------+-----------------------------------------------+
| P300 Single GEN2 (20-300 µL)    | * API v2.6 or higher: 92.86 µL/s              |
|                                 | * API v2.5 or lower: 46.43 µL/s               |
+---------------------------------+-----------------------------------------------+
| P1000 Single GEN2 (100-1000 µL) | * API v2.6 or higher: 274.7 µL/s              |
|                                 | * API v2.5 or lower: 137.35 µL/s              |
+---------------------------------+-----------------------------------------------+
| P20 Multi GEN2 (1-20 µL)        |    7.6 µL/s                                   |
+---------------------------------+-----------------------------------------------+
| P300 Multi GEN2 (20-300 µL)     |    94 µL/s                                    |
+---------------------------------+-----------------------------------------------+

> .. what is head speed, is it important/significant? applies to gantry movement or pipette? need to find out.

Reviewer: based on the following line, i think that's the speed that the gantry moves the pipette around at.

> Additionally, all OT-2 GEN2 pipettes have a default head speed of 400 mm/second and a well bottom clearance of 1mm for aspirate and dispense actions.

Reviewer: ```suggestion
Additionally, all OT-2 GEN2 pipettes have a default head speed of 400 mm/second and a well bottom clearance of 1 mm for aspirate and dispense actions.
```
🧐

Author: Committing and resolving.

The following table provides data on the default aspirate, dispense, and blow-out flow rates (in µL/s) for Flex pipettes.

+-------------------------+-------------+-----------------+-----------------+-----------------+
| Pipette Model           | Volume (µL) | Aspirate (µL/s) | Dispense (µL/s) | Blow-out (µL/s) |
+=========================+=============+=================+=================+=================+

> | Flex 1-Channel Pipette  | 1-50        | 8                                 | 4               |

Reviewer: wiiiiiide boi. let's list it twice: 8, 8, 4 and 160, 160, 80

Author: It is a wide load. We'll have to repeat the 160 and 80 three times because these have the flow rates for the 5-1000 uL pipettes:

- Flex 1-channel
- Flex 8-channel
- Flex 96-channel

With the revision, the only metrics that get the row spans are the volumes for the pipettes. Will fix and push.

+-------------------------+             +                                   +                 +
| Flex 8-Channel Pipette  |             |                                   |                 |
+-------------------------+-------------+-----------------------------------+-----------------+
| Flex 1-Channel Pipette  | 5-1000      | 160                               | 80              |
+-------------------------+             +                                   +                 +
| Flex 8-Channel Pipette  |             |                                   |                 |
+-------------------------+             +                                   +                 +
| Flex 96-Channel Pipette |             |                                   |                 |
+-------------------------+-------------+-----------------------------------+-----------------+

> Additionally, all Flex pipettes have a well bottom clearance of 1mm for aspirate and dispense actions.

Reviewer: ```suggestion
Additionally, all Flex pipettes have a well bottom clearance of 1 mm for aspirate and dispense actions.
```
😇

Author: Fixing locally. Will push.

OT-2 Pipette Flow Rates
-----------------------

> The following table provides data on the default aspirate, dispense, and blow-out flow rates (in µL/s) for OT-2 GEN2 pipettes. Because the flow rates are the same across types, we've consolidated the data into the "Flow Rates" column.

Reviewer: maybe `across all three actions` just so people don't think we're talking types of pipette.

Author: Yes, I thought this was wonk-a-doo weird. Will revise.

.. list-table::
    :header-rows: 1

    * - Pipette Model
      - Volume (µL)
      - Flow Rates (µL/s)
    * - P20 Single-Channel GEN2
      - 1-20
      - 
          * API v2.6 or higher: 7.56
          * API v2.5 or lower: 3.78
    * - P300 Single-Channel GEN2
      - 20-300
      - 
          * API v2.6 or higher: 92.86
          * API v2.5 or lower: 46.43
    * - P1000 Single-Channel GEN2
      - 100-1000
      -
          * API v2.6 or higher: 274.7
          * API v2.5 or lower: 137.35
    * - P20 Multi-Channel GEN2
      - 1-20
      - 7.6
    * - P300 Multi-Channel GEN2
      - 20-300
      - 94
 

> Additionally, all OT-2 GEN2 pipettes have a default head speed of 400 mm/second and a well bottom clearance of 1mm for aspirate and dispense actions.

Reviewer: `mm/s` and `1 mm` here…and then everything is neat and tidy!

Author: Fixing locally. Will push.
